### PR TITLE
Asset request pointer

### PR DESCRIPTION
**Pull request by sethalves**
- use SMART pointers for AssetRequest objects

### CODE REVIEW

File: libraries/networking/src/AssetClient.cpp [hunk 1]
```diff
@@ -175,8 +175,8 @@ RenameMappingRequest* AssetClient::createRenameMappingRequest(const AssetPath& o
     return request;
 }
 
-AssetRequest* AssetClient::createRequest(const AssetHash& hash) {
-    auto request = new AssetRequest(hash);
+AssetRequestPointer AssetClient::createRequest(const AssetHash& hash) {
+    auto request = AssetRequestPointer(new AssetRequest(hash));
```
Comment (Atlante45) [inline libraries/networking/src/AssetClient.cpp]: use `std::make_shared<>()`

File: libraries/networking/src/AssetRequest.h [hunk 2]
```diff
@@ -50,8 +55,13 @@ class AssetRequest : public QObject {
     const Error& getError() const { return _error; }
     QUrl getUrl() const { return ::getATPUrl(_hash); }
 
+    inline AssetRequestPointer getThisPointer() const {
+        return std::static_pointer_cast<AssetRequest>(std::const_pointer_cast<AssetRequest>(shared_from_this()));
+    }
```
Comment (Atlante45) [inline libraries/networking/src/AssetRequest.h]: We should probably have two version of this function, const/non const to avoid this const_cast.

**Comment by Atlante45** [comment]
Make sure to delete all the `request->deleteLater();` in the callbacks, otherwise you'll get double free crashes.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 670a1233aa95dfadd6b6e13686eef54b7eb799e2 available at: https://gist.github.com/5107614903b92ec467614f8672c478c4<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7679.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7679.dmg

File: libraries/networking/src/AssetRequest.h (hunk 2 above)
Comment (zzmp) [inline libraries/networking/src/AssetRequest.h]: Why inline?

File: libraries/networking/src/AssetRequest.h [hunk 3]
```diff
@@ -50,8 +55,13 @@ class AssetRequest : public QObject {
...
+
+
 signals:
-    void finished(AssetRequest* thisRequest);
+    void finished(AssetRequestPointer thisRequest);
```
Comment (zzmp) [inline libraries/networking/src/AssetRequest.h]: Why not store a `weak_ptr` and emit that, instead of inheriting from `std::enable_shared_from_this`?

File: libraries/networking/src/AssetResourceRequest.h [hunk 4]
```diff
@@ -36,7 +36,7 @@ private slots:
     void requestHash(const AssetHash& hash);
 
     GetMappingRequest* _assetMappingRequest { nullptr };
```
Comment (zzmp) [inline libraries/networking/src/AssetResourceRequest.h]: Is this OK being left as a raw pointer?

**Comment by hifi-gustavo** [comment]
Build console output for commit: 6f1aab9fe1669897e9dd5a8f546ca70a47528db9 available at: https://gist.github.com/5107614903b92ec467614f8672c478c4<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7679.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7679.dmg

**Comment by sethalves** [comment]
to test:
- get the default "home" content on your local machine.
- edit the cleanup robot and change its script to http://headache.hungry.com/~seth/hifi/reset.js
- click cleanup, watch the content be deleted and recreated repeatedly
- optionally, open windows' Task Manager and watch interface consume all RAM on the machine

**Comment by sethalves** [comment]
https://app.asana.com/0/26485730907942/107561861311879
https://app.asana.com/0/26485730907942/105483783922039
https://app.asana.com/0/26485730907942/110072858439679

**Comment by hifi-gustavo** [comment]
Build console output for commit: 6c5d07d4987ae68bd307c302fbbf3d5d6408dd57 available at: https://gist.github.com/5107614903b92ec467614f8672c478c4<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7679.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7679.dmg

File: libraries/networking/src/AssetRequest.h [hunk 5]
```diff
@@ -64,6 +70,9 @@ class AssetRequest : public QObject {
     int _numPendingRequests { 0 };
     MessageID _assetRequestID { AssetClient::INVALID_MESSAGE_ID };
     MessageID _assetInfoRequestID { AssetClient::INVALID_MESSAGE_ID };
+
+    AssetRequestWeakPointer _selfWeakPointer;
```
Comment (zzmp) [inline libraries/networking/src/AssetRequest.h]: Per coding standard, these should be `_self` and `setSelf`, without the word "Pointer".

File: libraries/networking/src/AssetRequest.cpp [hunk 6]
```diff
@@ -63,15 +69,16 @@ void AssetRequest::start() {
...
         _state = Finished;
-        emit finished(this);
+        emit finished(_selfWeakPointer);
         return;
     }
     
     _state = WaitingForInfo;
     
     auto assetClient = DependencyManager::get<AssetClient>();
+    auto saveMe = _selfWeakPointer.lock();
```
Comment (zzmp) [inline libraries/networking/src/AssetRequest.cpp]: This sounds like a desperate plea for help.

File: libraries/networking/src/AssetRequest.cpp [hunk 7]
```diff
@@ -63,15 +69,16 @@ void AssetRequest::start() {
...
     _assetInfoRequestID = assetClient->getAssetInfo(_hash,
-            [this](bool responseReceived, AssetServerError serverError, AssetInfo info) {
+            [this, saveMe](bool responseReceived, AssetServerError serverError, AssetInfo info) {
```
Comment (zzmp) [inline libraries/networking/src/AssetRequest.cpp]: You need to start by checking `saveMe`, or you risk access violations on these members.

File: libraries/networking/src/AssetRequest.cpp [hunk 8]
```diff
@@ -106,8 +113,9 @@ void AssetRequest::start() {
         int start = 0, end = _info.size;
         
         auto assetClient = DependencyManager::get<AssetClient>();
+        auto saveMe = _selfWeakPointer.lock();
         _assetRequestID = assetClient->getAsset(_hash, start, end,
-                [this, start, end](bool responseReceived, AssetServerError serverError, const QByteArray& data) {
+                [this, start, end, saveMe](bool responseReceived, AssetServerError serverError, const QByteArray& data) {
```
Comment (zzmp) [inline libraries/networking/src/AssetRequest.cpp]: You need to start by checking `saveMe`, or you risk access violations on these members.

File: libraries/networking/src/AssetRequest.cpp [hunk 9]
```diff
@@ -147,7 +155,7 @@ void AssetRequest::start() {
             }
             
             _state = Finished;
-            emit finished(this);
+            emit finished(_selfWeakPointer);
         }, [this](qint64 totalReceived, qint64 total) {
             emit progress(totalReceived, total);
```
Comment (zzmp) [inline libraries/networking/src/AssetRequest.cpp]: Can you do this without passing in `saveMe` and checking the validity before emitting from the object?

File: libraries/networking/src/AssetRequest.h [hunk 10]
```diff
@@ -12,6 +12,8 @@
 #ifndef hifi_AssetRequest_h
 #define hifi_AssetRequest_h
 
+#include <memory>
```
Comment (zzmp) [inline libraries/networking/src/AssetRequest.h]: Do you need this if you're just using Qt stuff?

File: libraries/networking/src/AssetResourceRequest.cpp [hunk 11]
```diff
@@ -107,8 +103,8 @@ void AssetResourceRequest::requestHash(const AssetHash& hash) {
     auto assetClient = DependencyManager::get<AssetClient>();
     _assetRequest = assetClient->createRequest(hash);
 
-    connect(_assetRequest, &AssetRequest::progress, this, &AssetResourceRequest::progress);
-    connect(_assetRequest, &AssetRequest::finished, this, [this](AssetRequest* req) {
+    connect(_assetRequest.data(), &AssetRequest::progress, this, &AssetResourceRequest::progress);
+    connect(_assetRequest.data(), &AssetRequest::finished, this, [this](AssetRequestPointer req) {
```
Comment (zzmp) [inline libraries/networking/src/AssetResourceRequest.cpp]: Why not make this another method, instead of a lambda? Seems cleaner.

File: libraries/networking/src/AssetRequest.h (hunk 3 above)
Comment (Atlante45) [inline libraries/networking/src/AssetRequest.h]: Why? That's exactly what `enable_shared_from_this` does.

Comment (Atlante45) [inline libraries/networking/src/AssetRequest.h]: Distilled for readability:
```cpp
template<class _Tp>
class enable_shared_from_this
{
    mutable weak_ptr<_Tp> __weak_this_;
protected:
    enable_shared_from_this() {}
    enable_shared_from_this(enable_shared_from_this const&) _NOEXCEPT {}
    enable_shared_from_this& operator=(enable_shared_from_this const&) {return *this;}
    ~enable_shared_from_this() {}
public:
    shared_ptr<_Tp> shared_from_this()
        {return shared_ptr<_Tp>(__weak_this_);}
    shared_ptr<_Tp const> shared_from_this() const
        {return shared_ptr<const _Tp>(__weak_this_);}

    template <class _Up> friend class shared_ptr;
};
```

Comment (zzmp) [inline libraries/networking/src/AssetRequest.h]: cppreference.com says:
> Note that prior to calling shared_from_this on an object t, there must be a std::shared_ptr that owns t.

Do you know what happens if that's not the case? Is it implementation-dependent?

Comment (Atlante45) [inline libraries/networking/src/AssetRequest.h]: Returns a null `std::weak_ptr`. But that's the point of this PR, AssetRequest should be managed by smart pointers.
If there's no `std::shared_ptr` owing it, then the object has been deleted already, or has been created "illegally".

File: libraries/networking/src/AssetRequest.cpp (hunk 7 above)
Comment (Atlante45) [inline libraries/networking/src/AssetRequest.cpp]: This is inside a method, it's totally safe to assume saveMe is valid otherwise we have another issue.

File: libraries/networking/src/AssetRequest.cpp (hunk 8 above)
Comment (Atlante45) [inline libraries/networking/src/AssetRequest.cpp]: same

File: libraries/networking/src/AssetRequest.cpp (hunk 7 above)
Comment (sethalves) [inline libraries/networking/src/AssetRequest.cpp]: I suspect that zzmp was thinking that saveMe was a weak pointer -- it's actually a shared pointer.

File: libraries/networking/src/AssetRequest.h (hunk 10 above)
Comment (sethalves) [inline libraries/networking/src/AssetRequest.h]: nope!

File: libraries/networking/src/AssetRequest.h (hunk 3 above)
Comment (zzmp) [inline libraries/networking/src/AssetRequest.h]: Cool. Thanks for explaining :). Guess I'm OK either way, then.

File: libraries/networking/src/AssetRequest.cpp (hunk 7 above)
Comment (zzmp) [inline libraries/networking/src/AssetRequest.cpp]: Correct.

File: libraries/networking/src/AssetRequest.cpp (hunk 9 above)
Comment (zzmp) [inline libraries/networking/src/AssetRequest.cpp]: retracted

**Comment by hifi-gustavo** [comment]
Build console output for commit: 17b4cdee3c57e26326333a05a288f2632e74e19d available at: https://gist.github.com/5107614903b92ec467614f8672c478c4<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7679.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7679.dmg

File: libraries/networking/src/AssetResourceRequest.h (hunk 4 above)
Comment (sethalves) [inline libraries/networking/src/AssetResourceRequest.h]: probably not, but I'm trying to limit the changes.

**Comment by hifi-gustavo** [comment]
Build console output for commit: b8f180964f758feb64086697315860972cecb671 available at: https://gist.github.com/5107614903b92ec467614f8672c478c4<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7679.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7679.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: c210fb21fd0e2d8938393a3e39d2c2676ec3b68a available at: https://gist.github.com/5107614903b92ec467614f8672c478c4<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7679.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7679.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 74e7b4e1edacf2d5b53ac69da52c35f28948eb41 available at: https://gist.github.com/5107614903b92ec467614f8672c478c4<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7679.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7679.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: ae46944c221e4807af25a38bda1c909afbd77af7 available at: https://gist.github.com/5107614903b92ec467614f8672c478c4<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7679.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7679.dmg

File: assignment-client/src/Agent.cpp [hunk 12]
```diff
@@ -245,8 +245,6 @@ void Agent::scriptRequestFinished() {
 
         setFinished(true);
     }
-
-    request->deleteLater();
```
Comment (Atlante45) [inline assignment-client/src/Agent.cpp]: ResourceRequest

File: interface/src/assets/ATPAssetMigrator.cpp [hunk 13]
```diff
@@ -239,8 +238,6 @@ void ATPAssetMigrator::setMappingFinished(SetMappingRequest* request) {
     }
 
     checkIfFinished();
-
-    request->deleteLater();
```
Comment (Atlante45) [inline interface/src/assets/ATPAssetMigrator.cpp]: SetMappingRequest

File: interface/src/scripting/AssetMappingsScriptingInterface.cpp [hunk 14]
```diff
@@ -50,8 +50,6 @@ void AssetMappingsScriptingInterface::setMapping(QString path, QString hash, QJS
             QJSValueList args { request->getErrorString(), request->getPath() };
             callback.call(args);
         }
-
-        request->deleteLater();
```
Comment (Atlante45) [inline interface/src/scripting/AssetMappingsScriptingInterface.cpp]: SetMappingRequest

File: interface/src/scripting/AssetMappingsScriptingInterface.cpp [hunk 15]
```diff
@@ -66,8 +64,6 @@ void AssetMappingsScriptingInterface::getMapping(QString path, QJSValue callback
             QJSValueList args { request->getErrorString() };
             callback.call(args);
         }
-
-        request->deleteLater();
```
Comment (Atlante45) [inline interface/src/scripting/AssetMappingsScriptingInterface.cpp]: GetMappingRequest

File: interface/src/scripting/AssetMappingsScriptingInterface.cpp [hunk 16]
```diff
@@ -136,8 +132,6 @@ void AssetMappingsScriptingInterface::deleteMappings(QStringList paths, QJSValue
...
```
Comment (Atlante45) [inline interface/src/scripting/AssetMappingsScriptingInterface.cpp]: DeleteMappingsRequest

File: interface/src/scripting/AssetMappingsScriptingInterface.cpp [hunk 17]
```diff
@@ -159,8 +153,6 @@ void AssetMappingsScriptingInterface::getAllMappings(QJSValue callback) {
             QJSValueList args { request->getErrorString(), map };
             callback.call(args);
         }
-
-        request->deleteLater();
```
Comment (Atlante45) [inline interface/src/scripting/AssetMappingsScriptingInterface.cpp]: GetAllMappingsRequest

File: interface/src/scripting/AssetMappingsScriptingInterface.cpp [hunk 18]
```diff
@@ -175,8 +167,6 @@ void AssetMappingsScriptingInterface::renameMapping(QString oldPath, QString new
             QJSValueList args { request->getErrorString() };
             callback.call(args);
         }
-
-        request->deleteLater();
```
Comment (Atlante45) [inline interface/src/scripting/AssetMappingsScriptingInterface.cpp]: RenameMappingRequest

File: libraries/controllers/src/controllers/UserInputMapper.cpp [hunk 19]
```diff
@@ -648,7 +648,6 @@ Mapping::Pointer UserInputMapper::newMapping(const QString& mappingName) {
 //        } else {
 //            qCWarning(controllers) << "Failed to load mapping url <" << jsonUrl << ">" << endl;
 //        }
-//        request->deleteLater();
```
Comment (Atlante45) [inline libraries/controllers/src/controllers/UserInputMapper.cpp]: ResourceRequest

File: libraries/networking/src/ResourceCache.cpp [hunk 20]
```diff
@@ -391,7 +410,6 @@ Resource::Resource(const QUrl& url, bool delayLoad) :
 Resource::~Resource() {
     if (_request) {
         _request->disconnect(this);
-        _request->deleteLater();
```
Comment (Atlante45) [inline libraries/networking/src/ResourceCache.cpp]: ResourceRequest

File: libraries/networking/src/ResourceCache.cpp [hunk 21]
```diff
@@ -444,7 +462,6 @@ void Resource::refresh() {
     }
     if (_request) {
         _request->disconnect(this);
-        _request->deleteLater();
```
Comment (Atlante45) [inline libraries/networking/src/ResourceCache.cpp]: ResourceRequest

**Comment by hifi-gustavo** [comment]
Build console output for commit: 04bf7513bfc65ec75e2947c47583da5d93e95efb available at: https://gist.github.com/9f10f503bb3ff22edd3edcd1f2cffdc3<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7679.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7679.dmg

File: libraries/networking/src/ResourceCache.cpp [hunk 22]
```diff
@@ -525,7 +542,6 @@ void Resource::reinsert() {
 void Resource::makeRequest() {
     if (_request) {
         _request->disconnect();
-        _request->deleteLater();
```
Comment (Atlante45) [inline libraries/networking/src/ResourceCache.cpp]: ResourceRequest

File: libraries/networking/src/ResourceCache.cpp [hunk 23]
```diff
@@ -605,7 +621,6 @@ void Resource::handleReplyFinished() {
     }
     
     _request->disconnect(this);
-    _request->deleteLater();
```
Comment (Atlante45) [inline libraries/networking/src/ResourceCache.cpp]: ResourceRequest

File: libraries/script-engine/src/BatchLoader.cpp [hunk 24]
```diff
@@ -51,7 +51,6 @@ void BatchLoader::start() {
                 _data.insert(url, QString());
                 qCDebug(scriptengine) << "Could not load" << url;
             }
-            request->deleteLater();
```
Comment (Atlante45) [inline libraries/script-engine/src/BatchLoader.cpp]: ResourceRequest

**Comment by hifi-gustavo** [comment]
Build console output for commit: 4894218f1c3d17dbad50223399fa8d55095299a6 available at: https://gist.github.com/5107614903b92ec467614f8672c478c4<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7679.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7679.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 5325e13a8742086f3547f60ea29daa6bb5ebae5d available at: https://gist.github.com/5107614903b92ec467614f8672c478c4<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7679.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7679.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: c7fd6b8f48f248d0aab973b3cea5d01d935d01a2 available at: https://gist.github.com/5107614903b92ec467614f8672c478c4<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7679.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7679.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 7582a16ee1dff3bd5e7e5484f1585d4eb0bc6ef8 available at: https://gist.github.com/5107614903b92ec467614f8672c478c4<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7679.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7679.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 784f129f17fc2fe32c2b105d4258954b58691f6b available at: https://gist.github.com/5107614903b92ec467614f8672c478c4<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7679.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7679.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 802a6ba53063786fb43776f4df130c003b957299 available at: https://gist.github.com/5107614903b92ec467614f8672c478c4<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7679.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7679.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 90e5a26e460d8cf6a4c6d0a54e23c1c968c29228 available at: https://gist.github.com/5107614903b92ec467614f8672c478c4<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7679.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7679.dmg

File: interface/src/Application.cpp [hunk 25]
```diff
@@ -1233,6 +1233,9 @@ void Application::cleanupBeforeQuit() {
 }
 
 Application::~Application() {
+    _entityClipboard->eraseAllOctreeElements();
```
Comment (zzmp) [inline interface/src/Application.cpp]: These changes belong in a different PR.

File: interface/src/avatar/AvatarManager.cpp [hunk 26]
```diff
@@ -77,7 +77,6 @@ AvatarManager::AvatarManager(QObject* parent) :
 }
 
 AvatarManager::~AvatarManager() {
-    _myAvatar->die();
```
Comment (zzmp) [inline interface/src/avatar/AvatarManager.cpp]: Why remove this? Also, this belongs in a separate PR.

File: interface/src/avatar/AvatarManager.cpp [hunk 27]
```diff
@@ -97,6 +96,12 @@ void AvatarManager::init() {
     scene->enqueuePendingChanges(pendingChanges);
 }
 
+void AvatarManager::clear() {
```
Comment (zzmp) [inline interface/src/avatar/AvatarManager.cpp]: Why not roll this into `die`, or remove `die` in favor of `clear`? Also, this belongs in a separate PR.

File: libraries/networking/src/AssetClient.cpp [hunk 28]
```diff
@@ -228,6 +228,7 @@ MessageID AssetClient::getAsset(const QString& hash, DataOffset start, DataOffse
 
         nodeList->sendPacket(std::move(packet), *assetServer);
 
+        // fill in GetAssetRequestData
```
Comment (zzmp) [inline libraries/networking/src/AssetClient.cpp]: Can you make this comment more clear?

File: libraries/networking/src/ResourceCache.cpp [hunk 29]
```diff
@@ -28,6 +28,99 @@
                            (((x) > (max)) ? (max) :\
                                             (x)))
 
+void ResourceCacheSharedItems::appendActiveRequest(QWeakPointer<Resource> resource) {
```
Comment (zzmp) [inline libraries/networking/src/ResourceCache.cpp]: This belongs in a separate PR.

File: libraries/networking/src/AssetResourceRequest.h (hunk 4 above)
Comment (zzmp) [inline libraries/networking/src/AssetResourceRequest.h]: Can you make an Asana issue noting this? It'll probably bite us later.

File: libraries/networking/src/AssetRequest.cpp [hunk 30]
```diff
@@ -127,27 +145,27 @@ void AssetRequest::start() {
...
+
             if (_error != NoError) {
                 qCWarning(asset_client) << "Got error retrieving asset" << _hash << "- error code" << _error;
             }
-            
+
             _state = Finished;
-            emit finished(this);
+            emit finished(thisPtr);
         }, [this](qint64 totalReceived, qint64 total) {
```
Comment (zzmp) [inline libraries/networking/src/AssetRequest.cpp]: Does this lambda need the same protection against this falling out of scope?

File: libraries/networking/src/AssetRequest.cpp [hunk 31]
```diff
@@ -51,27 +57,33 @@ void AssetRequest::start() {
...
-    
+
     _state = WaitingForInfo;
-    
+
     auto assetClient = DependencyManager::get<AssetClient>();
+    auto saveMe = _self;
     _assetInfoRequestID = assetClient->getAssetInfo(_hash,
-            [this](bool responseReceived, AssetServerError serverError, AssetInfo info) {
+            [this, saveMe](bool responseReceived, AssetServerError serverError, AssetInfo info) {
```
Comment (zzmp) [inline libraries/networking/src/AssetRequest.cpp]: Instead of passing both `this` and `saveMe`, can we just pass `saveMe` and use its members/methods? Are any methods/members private that require taking `this`?

**Comment by hifi-gustavo** [comment]
Build console output for commit: 0a717d7c0bcde1d5946391272dc664d9365a254b available at: https://gist.github.com/5107614903b92ec467614f8672c478c4<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7679.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7679.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 30549abe4ea37f29d22516c590e9b10a45d88239 available at: https://gist.github.com/5107614903b92ec467614f8672c478c4<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7679.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7679.dmg

**Comment by OmegaHeron** [comment]
20 Cycles of tidy up - no crashes.  This seems to fix my routine crash with tidybot in RC6+.

**Comment by OmegaHeron** [comment]
Adding general smoke test on Windows 7 - visit all the usual domains, place/remove objects - all good.

**Comment by sethalves** [comment]
@huffman (probably) found a better solution: #7731
